SwanCustomEnvironments/swancustomenvironments/scripts/makenv.sh

```diff
@@ -0,0 +1,292 @@
+ACCPY_BASE=$ACCPY_PATH/base
+ACCPY_ALL_VERSIONS_STR="?"
+PYTHON_PATH=$(which python)
+if [ -d "$ACCPY_PATH" ]; then
+    ACCPY_ALL_VERSIONS=$(ls -tr $ACCPY_BASE)
+    ACCPY_ALL_VERSIONS_STR=$(echo $ACCPY_ALL_VERSIONS | tr ' ' ', ')
+fi
+
+_log () {
+    if [ "$*" == "ERROR:"* ] || [ "$*" == "WARNING:"* ] || [ "${JUPYTER_DOCKER_STACKS_QUIET}" == "" ]; then
+        echo "$@"
+    fi
+}
+
+# Function for printing the help page
+print_help() {
+    _log "Usage: makenv --env/-e NAME --req/-r REQUIREMENTS [--accpy ACCPY_VERSION] [--clear/-c] [--help/-h]"
+    _log "Options:"
+    _log "  -e, --env NAME              Name of the custom virtual environment (mandatory)"
+    _log "  -r, --req REQUIREMENTS      Path to requirements.txt file or http link for a public repository (mandatory)"
+    _log "  -c, --clear                 Clear the current virtual environment, if it exists"
+    _log "  -h, --help                  Print this help page"
+    _log "  --accpy VERSION             Version of Acc-Py to be used (options: ${ACCPY_ALL_VERSIONS_STR})"
+}
+
+# --------------------------------------------------------------------------------------------
+
+# Parse command line arguments
+while [ $# -gt 0 ]; do
+    key="$1"
+    case $key in
+        --env|-e)
+            ENV_NAME=$2
+            shift
+            shift
+            ;;
+        --req|-r)
+            REQUIREMENTS=$2
+            shift
+            shift
+            ;;
+        --clear|-c)
+            CLEAR_ENV=--clear
+            shift
+            ;;
+        --help|-h)
+            print_help
+            exit 0
+            ;;
+        --accpy)
+            ACCPY_CUSTOM_VERSION=$2
+            shift
+            shift
+            ;;
+        *)
+            >&2 _log "ERROR: Invalid argument: $1" && _log
+            print_help
+            exit 1
+            ;;
+    esac
+done
+
+# --------------------------------------------------------------------------------------------
+
+ENV_PATH="/home/$USER/${ENV_NAME}"
+
+# Checks if a name for the environment is given
+if [ -z "$ENV_NAME" ]; then
+    >&2 _log "ERROR: No virtual environment name provided." && _log
+    print_help
+    exit 1
+fi
+
+# Checks if an environment with the same name was already created, if --clear is not passed
+if [ -d "$ENV_PATH" ] && [ -z "$CLEAR_ENV" ]; then
+    >&2 _log "ERROR: Virtual environment already exists."
+    exit 1
+fi
+
+# Checks if the provided Acc-Py version is valid
+if [ -n "$ACCPY_CUSTOM_VERSION" ] && [ ! -e "$ACCPY_BASE/$ACCPY_CUSTOM_VERSION" ]; then
+    >&2 _log "ERROR: Invalid Acc-Py version. Options: ${ACCPY_ALL_VERSIONS_STR}"
+    exit 1
+fi
+
+# Checks if a requirements file is given
+if [ -z "$REQUIREMENTS" ]; then
+    >&2 _log "ERROR: No requirements provided." && _log
+    print_help
+    exit 1
+# Checks if the provided requirements source is found
+elif [ -f $REQUIREMENTS ]; then
+    if [[ ${REQUIREMENTS##*.} != "txt" ]]; then
```

Comment:
This is to check if the path ends in .txt? Why not if it ends in `requirements.txt`?

Reply:
If you change the argument to be the path (EOS or git), you need to check here in which case you are (EOS or git), and try to find the requirements.txt differently in both cases, then error out if you can't find it.

Reply:
I made that verification to avoid RCEs attacks (or, at least, trying to). At that time, I didn't know if we would want to use other names for the requirements file, so I just accepted any name. That verification is only done with EOS requirements, if it is a Git repo, after cloned, we search for a requirements.txt file directly in the root folder of the repo. It would cause a lot of issues if we considered that it could be anywhere, because there could be multiple req. files, etc

Reply:
As discussed above in other threads, I would simplify this by assuming what you get in $REPOSITORY (new name for $REQUIREMENTS) is an EOS path or a git repo URL, then inside you look for the requirements.txt.